src/inc/til/u8u16convert.h

```diff
@@ -306,25 +91,160 @@ namespace til // Terminal Implementation Library. Also: "Today I Learned"
+            int len16{};
+            auto cursor8{ in.data() };
+            if (!!state.partials[3])
+            {
+                if (!len8) // only the partial code point will be converted and published
+                {
+                    len16 = MultiByteToWideChar(CP_UTF8, 0UL, state.partials.data(), gsl::narrow_cast<int>(state.partials[3]), out.data(), capa16);
+                    RETURN_HR_IF(E_UNEXPECTED, !len16);
+
+                    state.reset();
+                    out.resize(gsl::narrow_cast<size_t>(len16));
+                    return S_FALSE;
+                }
+
+                // determine the number of bytes of the code point indicated by the lead byte in the cache
+                const auto codePointLen{
+                    2 + // the code point consists of at least 2 bytes because if we cached data it will never be ASCII
+                    gsl::narrow_cast<int>((state.partials[0] & IsLead3) == IsLead3) + // add one if 3 high-order bits of the lead byte are set
+                    gsl::narrow_cast<int>((state.partials[0] & IsLead4) == IsLead4) // add another one if 4 high-order bits of the lead byte are set
+                };
+
+                const auto fetch8{ codePointLen - gsl::narrow_cast<int>(state.partials[3]) }; // number of bytes we need from the begin of the chunk to complete the code point
+                if (fetch8 > len8) // we still didn't get enough data to complete the code point, however this is not an error
+                {
+                    std::move(cursor8, cursor8 + len8, state.partials.data() + state.partials[3]);
+                    state.partials[3] += gsl::narrow_cast<char>(len8);
+                    out.clear();
+                    return S_OK;
+                }
+
+                std::move(cursor8, cursor8 + fetch8, state.partials.data() + state.partials[3]);
+                len16 = MultiByteToWideChar(CP_UTF8, 0UL, state.partials.data(), gsl::narrow_cast<int>(codePointLen), out.data(), capa16);
+                RETURN_HR_IF(E_UNEXPECTED, !len16);
+
+                state.reset();
+                capa16 -= len16;
+                len8 -= fetch8;
+                cursor8 += fetch8;
+            }
+
+            if (!!len8)
+            {
+                auto backIter{ cursor8 + len8 };
+                // If the last byte in the string was a byte belonging to a UTF-8 multi-byte character
+                if (!!(backIter[-1] & MaskAscii))
+                {
+                    // Check only up to 3 last bytes, if no Lead Byte is found then the byte before will be the Lead Byte and no partials are in the string
+                    const auto stopLen{ std::min(len8, 3) };
+                    for (auto sequenceLen{ 1 }; sequenceLen <= stopLen; ++sequenceLen)
+                    {
+                        --backIter;
+                        // If Lead Byte found
+                        if ((*backIter & MaskCont) > IsCont)

```

Comment:
`if ((*backIter & MaskCont) == MaskCont)` is fine. Although unnamed bitmasks are likely a better fit for that.

Reply:
Probably we can get rid of both `cmpMasks` and `cmpOperands`. Probably it's not worth to have the enum anymore in this case. Proper comments should be sufficient the. Let's see if @zadjii-msft would complain about it 😅

Reply:
Did it.